src/contexts/KeyringContextProvider.tsx

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges UI.
+//
+// Parity Bridges UI is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Parity Bridges UI is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges UI.  If not, see <http://www.gnu.org/licenses/>.
+
+//copied over from @substrate/context This needs to be updated.
+
+import { web3Accounts, web3Enable } from '@polkadot/extension-dapp';
+import type { KeyringPair } from '@polkadot/keyring/types';
+import keyring from '@polkadot/ui-keyring';
+import React, { useContext, useEffect, useReducer, useState } from 'react';
+
+import KeyringActions from '../actions/keyringActions';
+import keyringReducer from '../reducers/keyringReducer';
+import { KeyringContextType, KeyringStatuses } from '../types/keyringTypes';
+import { chainsConfigs } from '../util/substrateProviders';
+import { useSourceTarget } from './SourceTargetContextProvider';
+
+interface KeyringContextProviderProps {
+  children: React.ReactElement;
+}
+
+export const KeyringContext: React.Context<KeyringContextType> = React.createContext({} as KeyringContextType);
+
+export function useKeyringContext() {
+  return useContext(KeyringContext);
+}
+
+const INIT_STATE = {
+  keyring: null,
+  keyringStatus: null
+};
+
+export function KeyringContextProvider(props: KeyringContextProviderProps): React.ReactElement {
+  const { children = null } = props;
+  const { sourceChain } = useSourceTarget();
+  const { ss58Formats } = chainsConfigs;
+  const sourceChainSS58Format = ss58Formats[sourceChain];
+  const initState = { ...INIT_STATE };
+  const [state, dispatch] = useReducer(keyringReducer, initState);
+  const [keyringPairs, setKeyringPairs] = useState<Array<KeyringPair>>([]);
+  const [keyringPairsReady, setkeyringPairsReady] = useState(false);
```

Comment:
Why do we mix both the `useReducer` and `useState` here? It seems that information stored in both is tightly coupled. I think we should settle on either `useState` if the state is simple (and I think it is) or `useReducer` in case the state management is complicated (I think it isn't since we basically have one value & it's readiness).

Reply:
yes, i think the keyring reducer can be removed here. Actually my plan was to use this reducer as the account management reducer in general but i preferred to use the Account reducer that makes the keyring reducer to not have too much sense since it is just handling connection status updates that can live in a local state.